### PR TITLE
UDAs for function arguments

### PR DESCRIPTION
**Pull request by wilzbach**
#4783 revived from the dead.
See https://github.com/dlang/DIPs/pull/105 for a respective DIP.

Official spelling is ["user-defined attributes"](https://github.com/dlang/dlang.org/pull/2003#discussion_r157333463).

### CODE REVIEW

**Comment by wilzbach** [comment]
Jenkins failure looks unrelated: https://github.com/dlang/ci/pull/105

File: src/dmd/astbase.d [hunk 1]
```diff
@@ -1773,15 +1773,17 @@ struct ASTBase
...
         extern (D) alias ForeachDg = int delegate(size_t idx, Parameter param);
 
-        final extern (D) this(StorageClass storageClass, Type type, Identifier ident, Expression defaultArg)
+        final extern (D) this(StorageClass storageClass, Type type, Identifier ident, Expression defaultArg, UserAttributeDeclaration userAttribDecl)
         {
             this.storageClass = storageClass;
             this.type = type;
             this.ident = ident;
             this.defaultArg = defaultArg;
+            this.userAttribDecl = userAttribDecl;
```
Comment (wilzbach) [inline src/dmd/astbase.d]: @RazvanN7 do you have any plans on removing the giant code duplication in `astbase`?

**Comment by jacob-carlborg** [comment]
Will this work for template parameters?

**Comment by wilzbach** [comment]
> Will this work for template parameters?

Not this version. 
AFAICT is the proposed DIP also only about parameter attributes (and attaching attributes to enums).
It does work with template _d_ parameters - I just added test cases for that.

File: test/fail_compilation/udaparams.d [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+void vararg1(int a, @(10) ...);
+extern(C) void vararg2(int a, @(10) ...);
+
+void rhsuda(int a @(10));
+void rhsuda2(int @(10));
+void rhsuda3(int[] arr @(10) ...);
+
+void wrongAttr1(@safe int);
+void wrongAttr2(@safe void function());
+void wrongAttr3(@safe void delegate());
```
Comment (mathias-lang-sociomantic) [inline test/fail_compilation/udaparams.d]: That isn't going to be confusing at all :D

File: test/runnable/uda.d [hunk 3]
```diff
@@ -467,6 +467,113 @@ static assert(__traits(getAttributes, FileData11844.member)[0](new FileData11844
...
+{
+    void test17x(A)(@(23) A a)
+    {
+        static assert([__traits(getAttributes, a)] == [23]);
+    }
+    static assert([ParameterUDA!(0, test17x!int)] == [23]);
+}
+
+/************************************************/
+
```
Comment (mathias-lang-sociomantic) [inline test/runnable/uda.d]: Probably could use a test with a tuple as well

**Comment by jacob-carlborg** [comment]
> It does work with template d parameters - I just added test cases for that.

@wilzbach that do you mean with "template _d_ parameters"?

File: test/fail_compilation/udaparams.d (hunk 2 above)
Comment (skl131313) [reply test/fail_compilation/udaparams.d]: Those are invalid though attributes, so it shouldn't be too big of a problem.

File: src/dmd/astbase.d [hunk 4]
```diff
@@ -1773,15 +1773,17 @@ struct ASTBase
         Type type;
         Identifier ident;
         Expression defaultArg;
+        UserAttributeDeclaration userAttribDecl; // user defined attributes
 
         extern (D) alias ForeachDg = int delegate(size_t idx, Parameter param);
 
-        final extern (D) this(StorageClass storageClass, Type type, Identifier ident, Expression defaultArg)
+        final extern (D) this(StorageClass storageClass, Type type, Identifier ident, Expression defaultArg, UserAttributeDeclaration userAttribDecl)
```
Comment (skl131313) [inline src/dmd/astbase.d]:
```D
Expression defaultArg, UserAttributeDeclaration userAttribDecl = null)
```

Giving this a default value would save on a lot of code edits that just add a null as the last parameter. Not sure if default values are being avoided here on purpose or not.

File: test/fail_compilation/udaparams.d (hunk 2 above)
Comment (MetaLang) [reply test/fail_compilation/udaparams.d]: The following works currently:
```
void wrongAttr2(void function() @safe);
void wrongAttr3(void delegate() @safe);
```

And has the expected semantics. The question is, will it cause confusion that putting the `@safe` before the function literal type doesn't work?

**Comment by John-Colvin** [comment]
I'm trying to get a hold on the scope of this PR.
Can I do this?

```D
auto a(int x)(void function(@(x) int) f);
auto b = a((@(2) int){ });
```

**Comment by skl131313** [comment]
@John-Colvin 
Not I don't think that's possible with the current implementation, also not sure if that is desired.

File: src/dmd/astbase.d (hunk 4 above)
Comment (wilzbach) [reply src/dmd/astbase.d]: AFAICT is the prevalent coding style in the DMD source code to prefer explicitness.

**Comment by wilzbach** [comment]
So before we let this die again, how about making it available under `-transition=uda` (or `-dip=uda` or `-experimental=uda`)?
It won't get tested without being exposed and easily testable...

Other compilers have tons of experimental features that users can use to test upcoming features and even dmd has `-dip25`, `-dip1008` and `-dip1000` ...

E.g. rust has a lot more experiments going on: https://doc.rust-lang.org/stable/unstable-book/

**Comment by MetaLang** [comment]
We should only add such a switch for something that is almost certain to get into the language. Otherwise, people will begin using it in their code and it won't be able to removed later down the road, even if we want to.

**Comment by wilzbach** [comment]
Experimental switches would be like std.experimental - zero guarantees. No deprecations.

Remember that this addition has been submitted __three years ago__. Imho this level of progress for such tiny things (yes tiny compared to the other big fires D has) is really frustrating and from what I read so far from the responses in the State of D survey, many many other people feel the same. That's why they switched to Rust or kept using C++.

Experimental flags would also solve the chicken&egg problem of showing that it's useful without having a possibility to use it. We can of course also fork the compiler and implement new features there, but I doubt that such a split is a good idea.

**Comment by MetaLang** [comment]
> Experimental switches would be like std.experimental - zero guarantees. No deprecations.
I think we've found with std.experimental that telling people this just doesn't work. They will still use it in their code and will still get mad if we remove the feature and break their code.

> Remember that this addition has been submitted three years ago. Imho this level of progress for such tiny things (yes tiny compared to the other big fires D has) is really frustrating
I completely agree.

> Experimental flags would also solve the chicken&egg problem of showing that it's useful
I disagree.

**Comment by wilzbach** [comment]
> I disagree.

Well, do you have better ideas (except for forking the compiler?)

The only other idea that comes to my mind are feature branches. Martin used to popularize this concept (see e.g. http://nightlies.dlang.org), but they never gained traction because (1) they require manual builds which can only be done by him and (2) they feature branch still can get easily unmergeable into master (see eg. `newCTFE`) and (3) they aren't well-known and exposed.

**Comment by MetaLang** [comment]
> Do you have any better ideas

No I don't, which is frustrating, but my belief that experimental switches won't work is only becoming stronger; especially since people complained after (expected and announced) breaking changes to `std.experimental.allocator` were made. If a big "customer" such as Sociomantic, Weka, etc. start using the feature before it's fully fleshed out and we decide that it's not a good fit for D, good luck _ever_ getting it removed. If it ships with the compiler, people will use it, and they'll be upset when it breaks.

The only thing I can think of off the top of my head is compiler plugins that could be hosted on code.dlang.org. However, dmd currently doesn't support this kind of pluggable architecture, and I have no idea as to whether this could be made to work with LDC/GDC.

**Comment by jacob-carlborg** [comment]
> Well, do you have better ideas (except for forking the compiler?)

As Andrei said in another PR [1]:

"there's no reason for this limitation to exist and it breaks the rule of least astonishment. no dip needed, please make part of the changelog".

I would say we should go for it without a DIP.

[1] https://github.com/dlang/dmd/pull/6589#issuecomment-353998482

**Comment by MetaLang** [comment]
Thanks @jacob-carlborg, I didn't realize Andrei had already pre-approved this; that's great.

**Comment by jacob-carlborg** [comment]
>  I didn't realize Andrei had already pre-approved this; that's great.

@MetaLang he hasn't explicitly approved **this** change. But he thought it was fine to bypass the DIP process for another change: https://github.com/dlang/dmd/pull/6589#issuecomment-353998482.

**Comment by MetaLang** [comment]
Ha, I somehow thought the comment you linked to was on https://github.com/dlang/DIPs/pull/105. Considering that there _is_ a DIP for this feature, maybe we should wait until W&A render their verdict and then come back to this PR.

**Comment by skl131313** [comment]
That didn't turn out so great for that feature, it ended up getting removed. I don't think it is ready yet either. There aren't tests for failed compilation, and there isn't a test for lambda's either.

```D
// something like this I think ?

void test(alias func)()
{
    static if(is(typeof(func) PT == __parameters))
    {
        writeln(__traits(getAttributes, PT[0 .. 1]));
    }
}

test!((@("attribute") int param) => 0)();

```

**Comment by jacob-carlborg** [comment]
> That didn't turn out so great for that feature, it ended up getting removed.

True, but I don't think that was a useful addition to begin with.

> I don't think it is ready yet either. There aren't tests for failed compilation, and there isn't a test for lambda's either.

I didn't say we should merge it this second. Of course it needs to have the proper tests.

**Comment by JinShil** [comment]
> As Andrei said in another PR [1]:
>> "there's no reason for this limitation to exist and it breaks the rule of least astonishment. no dip needed, please make part of the changelog".

> I would say we should go for it without a DIP.

A little history on the DIP requirement:

1.  I posted to the forum asking for an official decision (https://forum.dlang.org/post/cltyrthdxkkfvuxqasqw@forum.dlang.org)
2.  @andralex said not to ask for official decisions on the forum (https://forum.dlang.org/post/ovhj5k$2131$1@digitalmars.com)
3.  I sent a private e-mail to @andralex and @WalterBright 
4.  @WalterBright replied saying a DIP is required.
5.  I posted to the forum stating that I had received an official decision https://forum.dlang.org/post/eiulsnlumpifcxwctnzj@forum.dlang.org
6.  I updated this PR with the "Pending DIP Approval" label

I consider the matter of whether a DIP is required for this or not a settled matter.  If you want the decision to be reconsidered, you'll have to ask @andralex and @WalterBright to do so.  Until then, a DIP is required.

**Comment by jacob-carlborg** [comment]
I just don’t think it’s fair that Walter and Andrei can bypass the DIP process while the rest of us have to struggle so much to get something through.

Also, as I’ve mentioned before, it doesn’t look like depreciations need to go through a similar process, which I think is weird.

**Comment by wilzbach** [comment]
> There aren't tests for failed compilation, and there isn't a test for lambda's either.

I'm aware, but as long as it's not foreseeable that this will be merged, I don't really feel inclined in investing more time in this. Though I don't want to stop other people from working with this (PR to this PR, directly pushing, just commenting with a diff, ...)

**Comment by andralex** [comment]
>I just don’t think it’s fair that Walter and Andrei can bypass the DIP process while the rest of us have to struggle so much to get something through.

That's... odd. Folks, no need to get agitated here. I'm trying to avoid bureaucracy where possible, not pull a Cambridge Analytica on a rainy country. Also note that the proposal and work don't originate from us...

**Comment by jacob-carlborg** [comment]
> I'm trying to avoid bureaucracy where possible

So why is bureaucracy needed for this change? Without this change UDA's are not consistent, not turtles all the way down.

**Comment by andralex** [comment]
>So why is bureaucracy needed for this change? Without this change UDA's are not consistent, not turtles all the way down.

My point exactly! So why are you reversing your position now? Are you running a Socratic discourse? Please clarify, thanks.

**Comment by jacob-carlborg** [comment]
> My point exactly! So why are you reversing your position now? Are you running a Socratic discourse? Please clarify, thanks.

I have not reversed my position. I don't think a DIP is required for this. Mike claims that Walter requires a DIP for this change, see his comment, point 4 https://github.com/dlang/dmd/pull/7576#issuecomment-374807297.

**Comment by JinShil** [comment]
@andralex What about #6161?  Is a DIP no longer required for that as well?

File: src/dmd/parse.d [hunk 5]
```diff
@@ -2896,6 +2915,28 @@ final class Parser(AST) : Lexer
...
+                            auto udad = new AST.UserAttributeDeclaration(udas, a);
+                            param.userAttribDecl = udad;
+                        }
+                        if (token.value == TOK.at)
+                        {
+                            AST.Expressions* exps = null;
+                            StorageClass stc2 = parseAttribute(&exps);
+                            if (stc2 == AST.STC.property || stc2 == AST.STC.nogc || stc2 == AST.STC.disable || stc2 == AST.STC.safe || stc2 == AST.STC.trusted || stc2 == AST.STC.system)
+                            {
+                                error("@%s attribute for function parameter is not supported", token.toChars());
```
Comment (JinShil) [inline src/dmd/parse.d]: Tick marks?

File: src/dmd/traits.d [hunk 6]
```diff
@@ -715,8 +717,12 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
         Identifier id;
         if (auto po = isParameter(o))
         {
+            if (!po.ident)
+            {
+                e.error("argument %s has no identifier", po.type.toChars());
```
Comment (JinShil) [inline src/dmd/traits.d]: Tick marks?

**Comment by andralex** [comment]
@JinShil I see a good DIP has been submitted, the plan is to approve that without too much paperwork.

**Comment by wilzbach** [comment]
Finally got around submitting the dlang.org PR: https://github.com/dlang/dlang.org/pull/2363

File: test/runnable/uda.d [hunk 7]
```diff
@@ -467,6 +467,128 @@ static assert(__traits(getAttributes, FileData11844.member)[0](new FileData11844
...
+void test14x(@Test14UDA1 int, @Test14UDA2("1") int, @test14uda3("2") int, @Test14UDA4!"3" int) {}
+void test14()
+{
+    static assert(is(ParameterUDA!(0, test14x)[0] == Test14UDA1));
+    static assert(ParameterUDA!(1, test14x)[0] == Test14UDA2("1"));
+    static assert(ParameterUDA!(2, test14x)[0] == test14uda3("2"));
+    static assert(is(ParameterUDA!(3, test14x)[0] == Test14UDA4!"3"));
+}
+
+void test15x(@(20) void delegate(int) @safe dg)
```
Comment (jacob-carlborg) [inline test/runnable/uda.d]: What about the parameters on delegates, are UDAs on those supported? Either way I think it's worth having a test case, showing either that it works or that it doesn't work.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

**Comment by IgorStepanov** [comment]
Ping. What status of that?

**Comment by jacob-carlborg** [comment]
> Ping. What status of that?

I asked for some more tests, which I don't think has been added. Not sure what else.

@wilzbach ping.

**Comment by wilzbach** [comment]
> I asked for some more tests, which I don't think has been added.

Sorry. I haven't found much time for this, but now managed to add these tests. See `test15y` and `test15z`.

> Not sure what else.

Same here.

File: test/runnable/uda.d (hunk 7 above)
Comment (wilzbach) [reply test/runnable/uda.d]: Added the tests below.

File: changelog/uda-function-arguments.dd [hunk 8]
```diff
@@ -0,0 +1,10 @@
+UDAs on function arguments are now supported
+
+User-defined attributes on function arguments behave analogous to existing UDAs:
+
+---
+void test(A)(@(22) A a)
+{
+    static assert([__traits(getAttributes, a)] == [22]);
+}
+---
```
Comment (jacob-carlborg) [inline changelog/uda-function-arguments.dd]: Perhaps include an example of how to get the UDAs of a parameter from outside of a function. Or should that be left as an exercise for the reader 😃.

**Comment by wilzbach** [comment]
> Lambdas, both with and without parentheses: alias a = @(3) b => 1 + 2 and alias b = (@(3) c, @(4) d) => 1 + 2 and usage like: map!(@(3) a => 1 + 2)

Only the latter was supported (i.e. with parentheses). I added support for `alias a = @(3) b => 1 + 2`.
Though I'm not sure that `alias a = @(3) b => 1 + 2 ` will be used in practice.
I added the tests to `test19`, but I'm not sure whether there's currently a way to query the attributes.

> Using a delegate with inferred parameter type: foo((@(3) a) { return 1 + 2; })

Added to `test20`.

> Template parameter: void foo(@(3) T)(T t)

That's not supported as of now. I'm not sure it's that useful, but if people run into use cases for it, it could always be added?
(I still added it to `fail_compilation`)

**Comment by jacob-carlborg** [comment]
> That's not supported as of now

I’m not asking for adding new functionality, just to add tests that shows it’s either compiles or fails to compile.

**Review by jacob-carlborg — APPROVED** [review]

File: changelog/uda-function-arguments.dd (hunk 8 above)
Comment (wilzbach) [reply changelog/uda-function-arguments.dd]: That's a bit more complicated and I think we will provide an interface in `std.traits` anyhow.

Comment (jacob-carlborg) [reply changelog/uda-function-arguments.dd]: Ok.

**Pull request merged**